src/language/tsLanguageService.ts

```diff
@@ -0,0 +1,215 @@
+          getLength: () => text.length,
+          getChangeRange: () => undefined,
+        };
+      },
+      getCurrentDirectory: () => '',
+      getDefaultLibFileName: () => GLOBAL_CONFIG_LIBRARY_NAME,
+      readFile: (path: string): string | undefined => {
+        if (path === currentTextDocument.uri) {
+          return currentTextDocument.getText();
+        }
+        return this._loadLibrary(path);
+      },
+      fileExists: (): boolean => false,
+      directoryExists: (): boolean => false,
+    };
+
+    // Create the language service files.
+    const jsLanguageService = ts.createLanguageService(host);
+
+    return {
+      // Return a new instance of the language service.
+      getLanguageService(jsDocument: TextDocument): ts.LanguageService {
+        currentTextDocument = jsDocument;
+        return jsLanguageService;
+      },
+      getCompilationSettings() {
+        return compilerOptions;
+      },
+      dispose() {
+        jsLanguageService.dispose();
+      },
+    };
+  }
+
+  doSignatureHelp(
+    document: TextDocument,
+    position: Position
+  ): Promise<SignatureHelp | null> {
+    const jsDocument = TextDocument.create(
+      document.uri,
+      'javascript',
+      document.version,
+      document.getText()
+    );
+    const jsLanguageService = this._host.getLanguageService(jsDocument);
+    const signHelp = jsLanguageService.getSignatureHelpItems(
+      jsDocument.uri,
+      jsDocument.offsetAt(position),
+      undefined
+    );
+
+    if (signHelp) {
+      const ret: SignatureHelp = {
+        activeSignature: signHelp.selectedItemIndex,
+        activeParameter: signHelp.argumentIndex,
+        signatures: [],
+      };
+      signHelp.items
+        .map((item) => {
+          const hasInt8Array = item.prefixDisplayParts.filter(
+            (prefix) => prefix.text !== 'Int8Array'
+          );
+          item.prefixDisplayParts = hasInt8Array;
+          return item;
+        })
+        .forEach((item) => {
+          const signature: SignatureInformation = {
+            label: '',
+            documentation: undefined,
+            parameters: [],
+          };
+
+          signature.label += ts.displayPartsToString(item.prefixDisplayParts);
+          item.parameters.forEach((p, i, a) => {
+            const label = ts.displayPartsToString(p.displayParts);
+            const parameter: ParameterInformation = {
+              label: label,
+              documentation: ts.displayPartsToString(p.documentation),
+            };
+            signature.label += label;
+            signature.parameters?.push(parameter);
+            if (i < a.length - 1) {
+              signature.label += ts.displayPartsToString(
+                item.separatorDisplayParts
+              );
+            }
+          });
+          signature.label += ts.displayPartsToString(item.suffixDisplayParts);
+          ret.signatures.push(signature);
+        });
```

Comment:
Maybe this is from editing leftovers, but it’s a bit odd that this code is using `.map()` with the semantics of `.forEach()` and `.forEach()` with the semantics of `.map()` … :slightly_smiling_face: This could probably be a `ret.signatures = signHelp.items.map((item) => { ...; return signature; })` call instead?